### ISSUE TITLE
Window close does not trigger save window state

### ISSUE THREAD

**Issue by Krajsnick** [atom/atom#13909]
### Description

Closing the Atom window does not save the window UI state, _read_ window size and position. To save the window state I have to **blur** (click outside) the window, then close it. Reopening the window after blurring it works.

According to the code in `window-event-handler.coffee` the window state should be stored both when a window closes and when blurred:
```coffee
handleWindowBlur: =>
    @atomEnvironment.storeWindowDimensions()

handleWindowBeforeunload: (event) =>
  @atomEnvironment.storeWindowDimensions()
```

### Steps to Reproduce

1. `atom .` to open project of your choice
2. Resize window, maintain window focus
3. Close window
4. Back to terminal, `atom .` again

Redo step **1-2** and before closing the window in step **3**, **blur** the window, click your desktop. Then close the window and continue with step **4**, the new window should open with the same dimensions as the window you just closed.

**Expected behavior:** The window you open in step **4** should be restored to the resized state, when you closed the window.

**Actual behavior:** The window you open in step **4** is restored back to the state before you resized, in step **2**.

### Versions

Atom    : 1.14.4
Electron: 1.3.13
Chrome  : 52.0.2743.82
Node    : 6.5.0
macOS 10.12.3

### Additional Information

I did try to start Atom in safe mode, and also with the `--clear-window-state` flag. Even deleting my `~/.atom/storage`.

You can also reload the window and see the same behaviour, instead of closing and re-opening a window. Resize Atom window then reload, watch it reset back to old dimensions, before reload. Watch the gif for a demo:

![window-state](https://cloud.githubusercontent.com/assets/979555/23480337/e966c98c-fec7-11e6-9ecd-91e23df89891.gif)

**Comment by rsese** [comment]
On 1.14.4 and macOS 10.12.3, I was able to reproduce this case:

> You can also reload the window and see the same behaviour, instead of closing and re-opening a window. Resize Atom window then reload, watch it reset back to old dimensions, before reload.

But not the clicking outside of Atom and then closing/opening a window case.

@Krajsnick when you say:

> before closing the window in step 3, blur the window, click your desktop. Then close the window

Do you mean that after clicking on the desktop, you just go back to the Atom window and click the x to close the Window?

**Comment by Krajsnick** [comment]
@rsese 
> Do you mean that after clicking on the desktop, you just go back to the Atom window and click the x to close the Window?

Yes, after having the window lose focus I just click the Atom editor pane again, then go for the red traffic light.

**Comment by rsese** [comment]
> These steps however, wont save window state for me - if I'm not blurring the window: between step 2 and 3.
> 
> atom . to open project of your choice
> Resize window, maintain window focus
> Close window
> Back to terminal, atom . again

Yep, I can reproduce this :+1:  Can you update the issue body with these new steps to reproduce?

I noticed a few similar but (I think) different issues that related specifically to maximized windows e.g. https://github.com/atom/atom/issues/10720 and https://github.com/atom/atom/issues/10181.

**Comment by Krajsnick** [comment]
@rsese Done! I have updated the procedure steps in my issue body.

**Comment by rsese** [comment]
:bow:

**Comment by 50Wliu** [comment]
Does it work if you quit Atom using Cmd+Q?

**Comment by rsese** [comment]
I had tried both Cmd+Q and just closing the window and in both cases, running Atom again with `atom .` both don't remember the window dimensions.

**Comment by J-qak** [comment]
It should be as simple as storing and restoring window sizes at window close. However I could not find the window close handler. Any luck? If the window close would also store window state, it would solve #10720 too.

**Comment by richardsimko** [comment]
I've fixed this issue in #15514 but I'm still looking for someone to have a peek at my PR and giving me some pointers to why my test might be failing.

**Comment by Krajsnick** [comment]
@richardsimko Snyggt!

**Comment by 50Wliu** [comment]
Fixed by #16295.
